src/ngraph/op/extractimagepatches.cpp

```diff
@@ -0,0 +1,124 @@
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//*****************************************************************************
+
+#include "ngraph/op/extractimagepatches.hpp"
+#include "ngraph/attribute_visitor.hpp"
+
+using namespace std;
+using namespace ngraph;
+
+// ExtractImagePatches v3
+
+constexpr NodeTypeInfo op::v3::ExtractImagePatches::type_info;
+
+op::v3::ExtractImagePatches::ExtractImagePatches(const Output<Node>& image,
+                                                 const Shape sizes,
+                                                 const Strides strides,
+                                                 const Shape rates,
+                                                 const PadType padding)
+    : Op({image})
+    , m_patch_sizes(sizes)
+    , m_patch_movement_strides(strides)
+    , m_patch_selection_rates(rates)
+    , m_padding(padding)
+{
+    constructor_validate_and_infer_types();
+}
+
+void op::v3::ExtractImagePatches::validate_and_infer_types()
+{
+    const PartialShape input_Pshape = get_input_partial_shape(0);
+
+    NODE_VALIDATION_CHECK(this,
+                          get_input_element_type(0).is_integral_number(),
+                          "input tensor must be an integral number.");
+    NODE_VALIDATION_CHECK(this, input_Pshape.rank() == 4, "input tensor must be 4D tensor.");
+
+    NODE_VALIDATION_CHECK(this,
+                          m_patch_sizes.size() == 2,
+                          "Attribute sizes should be in [size_rows, size_cols] format.");
+
+    NODE_VALIDATION_CHECK(this,
+                          m_patch_movement_strides.size() == 2,
+                          "Attribute strides should be in [stride_rows, stride_cols] format.");
+
+    NODE_VALIDATION_CHECK(this,
+                          m_patch_selection_rates.size() == 2,
+                          "Attribute rates should be in [rate_rows, rate_cols] format.");
+
+    NODE_VALIDATION_CHECK(
+        this,
+        m_padding == PadType::VALID || m_padding == PadType::SAME_LOWER ||
+            m_padding == PadType::SAME_UPPER,
+        "Attribute padding should be in either valid or same_lower or same_upper.");
+
+    if (input_Pshape.is_dynamic())
+    {
+        set_input_is_relevant_to_shape(0);
+        auto output_Pshape = PartialShape::dynamic(4);
+        set_output_type(0, get_input_element_type(0), output_Pshape);
+    }
+    else
+    {
+        Shape input_shape = get_input_shape(0); // as input shape is static
+        size_t out_rows((input_shape[2] - 1) / m_patch_movement_strides[0]);
```

Comment:
What padding type is this for? You overwrite padding below for VALID. The spec says for same upper/lower rows/cols match output.

If input_shape[2] were unsigned 0, subtracting 1 will give a very large number. I also think this gives the wrong values. If stride is 2, shape of 0->0, 1, 2->1, 3, 4->2, so (shape+stride-1)/stride

Also need to make sure strides are not 0.

Reply:
Thank you for discussing these thing with me, 
i will check following things:
so i will these 4 things:
   1. zero stride check
   2. stride > patch size check
   3. in_rows and in_cols zero check?
   4. Double check the dimension result with convolution and pooling validation_utils

Reply:
I updated the code.  none of the code doesn't have rates, so i used my tensorflow code to check the result of the dimensions